### PR TITLE
teleport,teleport.client: 4.2.11 -> 5.1.2

### PR DESCRIPTION
**Pull request by freezeboy**
###### Motivation for this change

Upstream update

And split the client `tsh` binary from the rest to reduce the size for "users" maybe servers could also be shrinked.

I would prefer someone to double check on their setup as it is quite a sensitive software.

Maybe could be also interesting to add a module and a test for it

I am also unsure if it is better for the default derivation to contain only the client or both the client and the server

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/teleport/default.nix [hunk 1]
```diff
@@ -19,26 +19,23 @@ buildGoPackage rec {
...
+  outputs = [ "out" "client" ];
 
+  buildTargets = [ "full" ];
   dontStrip = true;
 
-  meta = {
+  postInstall = ''
+    mkdir -p $client/bin
+    cp $out/bin/tsh $client/bin
+    strip $client/bin/tsh
```
Comment (SuperSandro2000) [inline pkgs/servers/teleport/default.nix:31]:
```suggestion
```

File: pkgs/servers/teleport/default.nix [hunk 2]
```diff
@@ -19,26 +19,23 @@ buildGoPackage rec {
...
-    rm -fr build/webassets.zip
-    cd $NIX_BUILD_TOP/go/bin
-    zip -q -A teleport
-    popd
-  '';
+  # Reduce closure size for client machines
+  outputs = [ "out" "client" ];
 
+  buildTargets = [ "full" ];
   dontStrip = true;
```
Comment (SuperSandro2000) [inline pkgs/servers/teleport/default.nix:26]:
```suggestion
```

File: pkgs/servers/teleport/default.nix [hunk 3]
```diff
@@ -19,26 +19,23 @@ buildGoPackage rec {
...
 
-  meta = {
+  postInstall = ''
+    mkdir -p $client/bin
+    cp $out/bin/tsh $client/bin
```
Comment (SuperSandro2000) [inline pkgs/servers/teleport/default.nix:30]:
```suggestion
    install -Dm755 -t tsh $client/bin
```

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which does not build all packages (e.g. lumo, tensorflow or pytorch)
If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 108487` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>teleport</li>
</ul>
</details>

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which does not build all packages (e.g. lumo, tensorflow or pytorch)
If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 108487` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>teleport</li>
</ul>
</details>

File: pkgs/servers/teleport/default.nix (hunk 2 above)
Comment (freezeboy) [reply pkgs/servers/teleport/default.nix:26]: I am not sure I can strip all binaries, the "teleport" binary contains some assests for the webapp and as the package was using `dontStrip` I wouldn't want to break the teleport server

**Comment by freezeboy** [comment]
I tried to strip everything, and run --help for the teleport app.
If someone with a service running on nixos can check that the binary it working, I'll be grateful

File: pkgs/servers/teleport/default.nix (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/servers/teleport/default.nix]: Please test if removing dontStrip is safe.

**Comment by Church-** [comment]
@freezeboy Nifty to see an update finally. I had given up getting it to work due to a bug with submodule cloning of teleports private e submodule a while back.

I have a PR for a module I've been meaning to refactor for ages, try and bang some work out on it this week to move it over to using @Infinisil's settings work.

**Comment by freezeboy** [comment]
Currently rebasing to merge with the update available in master

**Pull request merged**